### PR TITLE
Attr() cleanup; Shrink Entry interface; use plugin.CachedOp + plugin.CachedMetadata

### PR DESCRIPTION
**Pull request by ekinanp**

### CODE REVIEW

File: fuse/core.go [hunk 1]
```diff
@@ -51,6 +52,14 @@ func getIDs() (uint32, uint32) {
 
 var uid, gid = getIDs()
 
+func joinPath(parentPath string, e plugin.Entry) string {
```
Comment (ekinanp) [inline fuse/core.go]: Sorry to re-introduce this again. Figured it's better long-term, since it separates the FUSE concept of a path (which is filesystemy) from Wash's concept (which right now is filesystemy, but may not be in the future depending on how annoying the issue described in https://github.com/puppetlabs/wash/issues/156 is)

File: plugin/aws/ec2InstanceConsoleOutput.go [hunk 2]
```diff
@@ -29,30 +26,29 @@ func newEC2InstanceConsoleOutput(inst *ec2Instance, latest bool) *ec2InstanceCon
...
+	output, err := cl.inst.cachedConsoleOutput(ctx, cl.latest)
 	if err != nil {
-		return nil, err
+		return plugin.Attributes{}, err
 	}
 
-	content, err := base64.StdEncoding.DecodeString(awsSDK.StringValue(resp.Output))
+	cl.Ctime = output.mtime
+	attr, _ := cl.EntryBase.Attr(ctx)
+	attr.Size = uint64(len(output.content))
```
Comment (ekinanp) [inline plugin/aws/ec2InstanceConsoleOutput.go]: Do we even need `Size` ? Could this be calculated from `SizedReader` instead (and have EntryBase mark the Size attribute as unknown by default)? Am leaning towards that. This way, it's clear that an S3 object's Ctime is its Mtime (which is precisely how S3 models its objects -- they're atomic blobs).

File: plugin/aws/ec2InstanceMetadataJSON.go [hunk 3]
```diff
@@ -23,7 +23,7 @@ func newEC2InstanceMetadataJSON(inst *ec2Instance) *ec2InstanceMetadataJSON {
 }
 
 func (im *ec2InstanceMetadataJSON) Open(ctx context.Context) (plugin.SizedReader, error) {
```
Comment (ekinanp) [inline plugin/aws/ec2InstanceMetadataJSON.go]: Could we turn off caching for Open here? How much overhead would prettifying the metadata impose on performance?

File: plugin/entryBase.go [hunk 4]
```diff
@@ -67,9 +69,19 @@ func (e *EntryBase) ID() string {
...
+// are set to this value.
+//
+// Entries can override Attr if they'd like to return a different
+// set of attributes.
 func (e *EntryBase) Attr(ctx context.Context) (Attributes, error) {
-	return e.attr, nil
+	return Attributes{
+		Ctime: e.Ctime,
+		Mtime: e.Ctime,
+		Atime: e.Ctime,
```
Comment (ekinanp) [inline plugin/entryBase.go]: Should plugin authors worry about `Atime` ?  It seems like if you do anything with any entry in Wash, you're "accessing" it so `Atime` should always be `time.Now()`. Hm, maybe for sshfs-type files it could be useful.

File: plugin/types.go [hunk 5]
```diff
@@ -39,10 +39,10 @@ import (
 // creating your plugin objects.
 type Entry interface {
 	Name() string
-	ID() string
 	Attr(ctx context.Context) (Attributes, error)
-	getTTLOf(op actionOpCode) time.Duration
+	id() string
```
Comment (ekinanp) [inline plugin/types.go]: Unexporting the id also prevents plugin authors from overwriting it.

File: plugin/entryBase.go (hunk 4 above)
Comment (MikaelSmith) [reply plugin/entryBase.go]: `Atime` generally means when it was "read". For actual files it makes sense. For other things I'm not sure.

Comment (ekinanp) [reply plugin/entryBase.go]: Oh ok. Might as well keep it then.

File: plugin/aws/ec2Instance.go [hunk 6]
```diff
@@ -64,16 +44,15 @@ const (
...
 	ec2Instance := &ec2Instance{
 		EntryBase:        plugin.NewEntry(ID),
 		session:          session,
 		client:           client,
 		ssmClient:        ssm.New(session),
 		cloudwatchClient: cloudwatchlogs.New(session),
-		attr:             attr,
 	}
-	ec2Instance.TurnOffCachingFor(plugin.List)
+	ec2Instance.TurnOffCaching()
```
Comment (MikaelSmith) [inline plugin/aws/ec2Instance.go]: I'd definitely like a comment here. `TurnOffCaching` doesn't really describe what's happening anymore either; it's more like `DisableDefaultCaching` (because it only affects the ops that are native to the plugin package).

File: plugin/aws/ec2InstanceMetadataJSON.go (hunk 3 above)
Comment (MikaelSmith) [reply plugin/aws/ec2InstanceMetadataJSON.go]: I think we could, it should be very minimal overhead.

File: plugin/aws/s3Object.go [hunk 7]
```diff
@@ -36,49 +36,62 @@ func newS3Object(bucket string, key string, client *s3Client.S3) *s3Object {
...
 }
 
 func (o *s3Object) Attr(ctx context.Context) (plugin.Attributes, error) {
-	resp, err := o.cachedHeadObject(ctx)
+	result, err := o.cachedHeadObject(ctx)
 	if err != nil {
 		return plugin.Attributes{}, nil
 	}
 
-	size := awsSDK.Int64Value(resp.ContentLength)
```
Comment (MikaelSmith) [inline plugin/aws/s3Object.go]: Why did you move this within the cached operation?

File: plugin/aws/ec2Instance.go (hunk 6 above)
Comment (ekinanp) [reply plugin/aws/ec2Instance.go]: Agreed. I think `TurnOffCachingFor` => `DisableCachingFor`, `TurnOffCaching` => `DisableDefaultCaching` makes more sense.

File: plugin/entryBase.go [hunk 8]
```diff
@@ -24,11 +24,13 @@ var actionOpCodeToNameMap = [3]string{"List", "Open", "Metadata"}
 // EntryBase implements Entry, making it easy to create new entries.
 // You should use plugin.NewEntry to create new EntryBase objects.
 type EntryBase struct {
-	name string
+	// Ctime specifies the entry's creation time. See Attr for more
+	// details on how this is used.
+	Ctime time.Time
```
[the remark this answers is not in the document]
Comment (ekinanp) [reply plugin/entryBase.go]: Yeah, I think you're right. I'll revert it

File: plugin/aws/s3Object.go (hunk 7 above)
Comment (ekinanp) [reply plugin/aws/s3Object.go]: So there wouldn't be a race condition on `Ctime` (since I'm setting it then using EntryBase's `Attr`). But that indicates `Ctime` is a bad idea, so I'll go ahead and remove it then take this out of the cached operation.

File: plugin/entryBase.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ekinanp) [reply plugin/entryBase.go]: It feels weird to me b/c it's early taxonomy that doesn't give a lot of value and is an extra thing plugin authors have to keep in mind. It's much simpler to just say "Implement Attr if your entry has attributes".

[the remark this answers is not in the document]
Comment (MikaelSmith) [reply plugin/entryBase.go]: It does make some plugins cleaner. And it should never change. It feels like a very pragmatic improvement.

File: plugin/aws/s3Object.go (hunk 7 above)
Comment (MikaelSmith) [reply plugin/aws/s3Object.go]: Ah, you also want to be careful calling CachedOp from within another CachedOp. They could end up using the same mutex and deadlock.

File: plugin/entryBase.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (ekinanp) [reply plugin/entryBase.go]: Hmm, I think this could be less weird if we reinforced the `it should never change` fact. I'll explore that.

**Comment by ekinanp** [comment]
Accidentally pushed up the caching updates. Still addressing the remaining feedback in the PR.

File: plugin/cache.go [hunk 9]
```diff
@@ -87,6 +88,10 @@ type opFunc func() (interface{}, error)
 // For example, CachedOp could be useful to cache an API request whose response
 // lets you implement Attr() and Metadata() for the given entry.
 func CachedOp(opName string, entry Entry, ttl time.Duration, op opFunc) (interface{}, error) {
+	if !opNameRegex.MatchString(opName) {
```
Comment (MikaelSmith) [inline plugin/cache.go]: This isn't quite right. `opNameRegex` allows anything to follow after the regex, it needs to end with `$`.

File: plugin/cache.go [hunk 10]
```diff
@@ -49,8 +49,8 @@ func UnsetTestCache() {
 	cache = nil
 }
 
-var opNameRegex = regexp.MustCompile("^[a-zA-Z]+")
-var opQualifier = opNameRegex.String() + "::"
+var opNameRegex = regexp.MustCompile("^[a-zA-Z]+$")
+var opQualifier = "^[a-zA-Z]+::"
```
Comment (MikaelSmith) [inline plugin/cache.go]: This could be a `const`.

Comment (MikaelSmith) [reply plugin/cache.go]:
```suggestion
const opQualifier = "^[a-zA-Z]+::"
```
